HW6 is completed

albkharisov opened the pull request:
## Домашнее задание №6 «Пайплайн»

### Чек-лист студента ([Что это?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0))
- [x] Я перечитал задание после решения. [Зачем перечитывать задание?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D0%BF%D0%B5%D1%80%D0%B5%D1%87%D0%B8%D1%82%D1%8B%D0%B2%D0%B0%D1%82%D1%8C-%D0%B7%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5)
- [x] Я запустил `go mod tidy`.
- [x] Я удалил `.sync` файл. [Зачем его удалять?](https://github.com/OtusGolang/home_work/wiki/%5B%D0%A1%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0%D0%BC%5D-%D0%9F%D1%80%D0%BE%D1%86%D0%B5%D1%81%D1%81-%D1%81%D0%B4%D0%B0%D1%87%D0%B8-%D0%94%D0%97#user-content-%D0%92%D0%B0%D1%80%D0%B8%D0%B0%D0%BD%D1%82-2)
- [x] Я сравнил объем кода с объемом авторского решения. [Зачем сверять объем кода?](https://github.com/OtusGolang/home_work/wiki/%D0%9A%D0%BE%D0%BC%D0%BC%D0%B5%D0%BD%D1%82%D0%B0%D1%80%D0%B8%D0%B8-%D0%BA-%D1%87%D0%B5%D0%BA-%D0%BB%D0%B8%D1%81%D1%82%D1%83-%D1%81%D1%82%D1%83%D0%B4%D0%B5%D0%BD%D1%82%D0%B0#user-content-%D0%97%D0%B0%D1%87%D0%B5%D0%BC-%D1%81%D0%B2%D0%B5%D1%80%D1%8F%D1%82%D1%8C-%D0%BE%D0%B1%D1%8A%D0%B5%D0%BC-%D0%BA%D0%BE%D0%B4%D0%B0)
  - pipeline.go ~55 строк

### Критерии оценки
- [x] CI-пайплайн зелёный - 5 баллов
- [x] Добавлены новые юнит-тесты - до 2 баллов
- [x] Понятность и чистота кода - до 3 баллов

#### Зачёт от 7 баллов

albkharisov committed:
HW6 is completed

albkharisov committed:
Merge branch 'master' into hw06_pipeline_execution

agneum commented on hw06_pipeline_execution/pipeline.go:
```diff
@@ -8,7 +8,37 @@ type (
 func ExecutePipeline(in In, done In, stages ...Stage) Out {
-	// Place your code here.
-	return nil
+	dataCh := in
```
Можно обернуть `in` во `wrapChannel`, чтобы была возможность остановить пайплайн, если `stages` пуст

albkharisov replied:
Спасибо! Хорошее замечание.

Забавно. Я написал [тест empty stages](https://github.com/albkharisov/hw/pull/8/files#diff-edf6b4bf550619330e874d4dd0417073a266a0be5e72d20057e1c1f8ee47542cR94), который сразу закрывал канал, но ожидал от него другого поведения - что все значения просто пройдут насквозь и ничего не будет остановлено.

Поправил тест и поведение.

Нашёл ещё пару ошибок (приоритет done channel надо выставлять выше) пока правил тест. Проявляется если запустить тест несколько раз.

agneum reviewed (approved):
Спасибо за выполненное задание. Хорошая работа
Итого: 10 баллов из 10

Просьба при создании Pull Request использовать шаблон. Подробнее в инструкции по сдаче ДЗ

albkharisov committed:
Fix 'no stages' case

Allow closing pipeline by done channel even if no stages provided.
Previous logic passed all values through despite of done
channel.
Correct test cases, which expected wrong behaviour.

albkharisov committed:
Merge branch 'master' into hw06_pipeline_execution
